File: android/jvm-tests/src/test/kotlin/PartnerTest.kt
import android.content.Context
import hu.breaker.app.core.AppState
import hu.breaker.app.core.Blocklist
import hu.breaker.app.core.BreakerStore
import hu.breaker.app.core.ChallengeEngine
import hu.breaker.app.core.ChallengeEngine.Kind
import hu.breaker.app.core.ChallengeEngine.Step
import hu.breaker.app.core.Focus
import hu.breaker.app.core.FocusSync
import hu.breaker.app.core.PartnerLogic
import hu.breaker.app.core.Referee
import hu.breaker.app.core.Site
import hu.breaker.app.core.SyncClient
import hu.breaker.app.core.SyncRevisions
import org.json.JSONObject
import java.io.File
import kotlin.test.BeforeTest
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith
import kotlin.test.assertFalse
import kotlin.test.assertNotEquals
import kotlin.test.assertNotNull
import kotlin.test.assertNull
import kotlin.test.assertTrue

/**
 * Párban zárolás a Kotlin tükrön — a desktop/test/partner.test.ts esetei. A
 * lenyomat a `fixtures/partner-hash.json`-nal mérve: a gépen felvett
 * megbízottnak a telefonon is stimmelnie kell.
 */
class PartnerTest {

    private val now = 1_700_000_000_000L
    private val phrase = "alma bogrács cinege délután"

    @BeforeTest fun reset() {
        BreakerStore.init(Context())
        BreakerStore.mutate { AppState() }
        BreakerStore.saveLastTick(0)
    }

    private fun fixtureFile(): File {
        var dir: File? = File(".").absoluteFile
        while (dir != null) {
            val f = File(dir, "fixtures/partner-hash.json")
            if (f.isFile) return f
            dir = dir.parentFile
        }
        error("fixtures/partner-hash.json nincs meg a tároló gyökerében")
    }

    @Test fun `a jelmondat kanonikus alakja - kis-nagybetu, szokozok, NFKC nem szamit`() {
        assertEquals(phrase, PartnerLogic.normalizePhrase("  Alma  BOGRÁCS\tcinege\n délután "))
        // Bontott ékezet (a + kombináló vessző) ugyanaz, mint az összetett á.
        assertEquals(PartnerLogic.normalizePhrase("álma"), PartnerLogic.normalizePhrase("a" + Char(0x0301) + "lma"))
        assertEquals("", PartnerLogic.normalizePhrase("   "))
        assertEquals("Anna Kovács", PartnerLogic.normalizePartnerName("  Anna   Kovács "))
        assertNull(PartnerLogic.normalizePartnerName(""))
        assertNull(PartnerLogic.normalizePartnerName(null))
        assertEquals(PartnerLogic.MAX_PARTNER_NAME, PartnerLogic.normalizePartnerName("x".repeat(100))!!.length)
    }

    @Test fun `a lenyomat a gepevel egyezik - fixtures partner-hash json`() {
        val fx = JSONObject(fixtureFile().readText())
        val salt = fx.getString("salt")
        assertEquals(fx.getString("hash"), PartnerLogic.hashPhrase(fx.getString("phrase"), salt), "ugyanaz a jelmondat, ugyanaz a lenyomat — mint a gépen")
        assertEquals(fx.getString("hash"), PartnerLogic.hashPhrase(" Alma bogrács CINEGE délután", salt), "a kanonikus alak számít")
        assertNotEquals(fx.getString("hash"), PartnerLogic.hashPhrase(fx.getString("wrong"), salt))
        val lock = PartnerLogic.PartnerLock("Anna", salt, fx.getString("hash"), 1L)
        assertTrue(PartnerLogic.verify(lock, "ALMA  bogrács cinege délután "))
        assertFalse(PartnerLogic.verify(lock, fx.getString("wrong")))
        assertFalse(PartnerLogic.verify(lock, ""))
        val made = PartnerLogic.makeLock("Anna", phrase, now)
        assertTrue(PartnerLogic.verify(made, phrase))
        assertNotEquals(salt, made.salt, "friss só minden felvételnél")
        assertNull(PartnerLogic.normalizeLock("Anna", "rövid", fx.getString("hash"), 1L))
        assertNull(PartnerLogic.normalizeLock("", salt, fx.getString("hash"), 1L))
        assertEquals(lock, PartnerLogic.normalizeLock("Anna", salt, fx.getString("hash"), 1L))
    }

    @Test fun `fesules - a jel dont, azonos jelnel a beallitott, es a korabban felvett`() {
        val a = PartnerLogic.PartnerLock("Anna", "A".repeat(24), "B".repeat(44), 100)
        val b = PartnerLogic.PartnerLock("Béla", "C".repeat(24), "D".repeat(44), 200)
        assertNull(PartnerLogic.merge(3, a, 5, null), "a nagyobb jelű levétel átmegy")
        assertNull(PartnerLogic.merge(5, null, 3, a), "a helyi, nagyobb jelű levétel marad")
        assertEquals(a, PartnerLogic.merge(3, a, 3, null), "azonos jelnél a beállított nyer")
        assertEquals(b, PartnerLogic.merge(3, null, 3, b))
        assertEquals(a, PartnerLogic.merge(3, b, 3, a), "mindkettő beállítva: a korábban felvett")
        assertEquals(b, PartnerLogic.merge(2, a, 3, b), "nagyobb jel: a másik megbízott")

        val base = FocusSync.SyncFocus(packs = emptyList(), run = null, log = emptyList(), rev = 0, updatedAt = 0, updatedBy = "dev")
        val local = base.copy(partner = a, partnerRev = 3, rev = 3)
        val removed = base.copy(partnerRev = 5, rev = 5, updatedBy = "other")
        val merged = FocusSync.merge(local, removed)
        assertNull(merged.partner)
        assertEquals(5, merged.partnerRev)
        // A régi kliens blobja (jel nélkül) sosem viszi el a megbízottat.
        val old = base.copy(rev = 9, updatedAt = 999, updatedBy = "old")
        assertEquals(a, FocusSync.merge(local, old).partner)
        assertFalse(FocusSync.same(local, local.copy(partner = b)), "a megbízott cseréje különbség: fel kell tölteni")
    }

    @Test fun `drot-alak - a megbizott JSON-ja oda-vissza, a rossz alaku kiesik`() {
        val a = PartnerLogic.PartnerLock("Anna", "A".repeat(24), "B".repeat(44), 100)
        assertEquals(a, SyncClient.partnerFromJson(SyncClient.partnerToJson(a)))
        assertNull(SyncClient.partnerFromJson(JSONObject().put("name", "X").put("salt", "rövid").put("hash", "rövid")))
        assertNull(SyncClient.partnerFromJson(null))
        val f = SyncClient.focusFromJson(
            JSONObject().put("packs", org.json.JSONArray()).put("rev", 4).put("partnerRev", 99)
                .put("partner", SyncClient.partnerToJson(a)).toString(),
            "dev", null,
        )
        assertEquals(a, f.partner)
        assertNull(f.partnerRev, "a jel legfeljebb a blob rev-je")
    }

    // ------------------------------------------------------------------ a bíró

    private fun addSite(domain: String): String {
        val id = BreakerStore.newId("site")
        BreakerStore.mutate { s ->
            s.copy(sites = s.sites + Site(
                id = id, domain = domain, hostnames = Blocklist.expandHostnames(domain, false),
                addedAt = now, pauseUntil = null, pendingDeleteAt = null,
            ))
        }
        return id
    }

    private fun currentStep(): Step {
        val s = BreakerStore.state.value.session!!
        return s.steps[s.stepIndex]
    }

    private fun solve(step: Step): String = when (step) {
        is Step.Transcribe -> step.text
        is Step.MathChain -> step.problems[step.pos].a.toString()
        is Step.Memory -> {
            BreakerStore.mutate { s ->
                val ses = s.session!!
                val steps = ses.steps.toMutableList()
                steps[ses.stepIndex] = step.copy(armedAt = now - step.showMs - step.waitMs - 1000)
                s.copy(session = ses.copy(steps = steps))
            }
            step.code
        }
        is Step.Reverse -> step.text.reversed()
        is Step.Delay -> error("a várakozást átvenni kell")
        is Step.Partner -> error("a megbízott lépése a jelmondat")
    }

    /** Végigviszi a kísérletet a megbízott lépéséig — a várakozást is átveszi. */
    private fun solveUntilPartner(id: String) {
        var guard = 0
        while (BreakerStore.state.value.session != null && guard++ < 200) {
            when (val step = currentStep()) {
                is Step.Partner -> return
                is Step.Delay -> Referee.claimDelay(id, (step.claimableAt ?: now) + 1)
                else -> Referee.submitAnswer(id, solve(step), now)
            }
        }
        error("a kísérlet elfogyott a megbízott lépése előtt")
    }

    @Test fun `megbizottal a terv utolso lepese az o jelmondata - a varakozas utan`() {
        val siteId = addSite("youtube.com")
        val setup = Referee.setPartner("  Anna ", now)
        assertEquals("Anna", setup.name)
        assertEquals(PartnerLogic.PARTNER_PHRASE_WORDS, setup.phrase.split(" ").size, "négy szó")
        assertEquals(setup.phrase, setup.phrase.lowercase())
        assertEquals("Anna", BreakerStore.state.value.partner?.name)
        assertFailsWith<Referee.RefereeException> { Referee.setPartner("Béla", now) }

        val ses = Referee.startSession(Kind.PAUSE, siteId, 15, now)
        assertTrue(ses.steps.last() is Step.Partner)
        assertTrue(ses.steps[ses.steps.size - 2] is Step.Delay)
        assertEquals("Anna", (ses.steps.last() as Step.Partner).name)

        solveUntilPartner(ses.id)
        // Rossz jelmondat: nem sorsol újat, csak számol; a plafonnál a kísérlet elszáll.
        repeat(PartnerLogic.MAX_PARTNER_TRIES - 1) {
            val r = Referee.submitAnswer(ses.id, "alma bogrács cinege este", now)
            assertFalse(r.accepted)
            assertTrue(r.message!!.contains("Nem ez a jelmondat"))
            assertNotNull(BreakerStore.state.value.session, "a kísérlet még él")
        }
        val last = Referee.submitAnswer(ses.id, "megint rossz", now)
        assertFalse(last.accepted)
        assertTrue(last.message!!.contains("elölről"))
        assertNull(BreakerStore.state.value.session, "a plafonnál elszállt")
        assertNull(BreakerStore.state.value.sites[0].pauseUntil, "feloldás nem történt")

        // Újra: minden lépés elölről, és a jó jelmondat a végén feloldja.
        val again = Referee.startSession(Kind.PAUSE, siteId, 15, now + 1000)
        solveUntilPartner(again.id)
        val r = Referee.submitAnswer(again.id, " ${setup.phrase.uppercase()} ", now + 1000)
        assertTrue(r.accepted)
        assertTrue(r.sessionDone)
        assertNotNull(BreakerStore.state.value.sites[0].pauseUntil, "a szünet elindult")
    }

    @Test fun `a feladott kiserlet kombinacioja nem tartalmazza a megbizott lepeset`() {
        val siteId = addSite("youtube.com")
        Referee.setPartner("Anna", now)
        val ses = Referee.startSession(Kind.PAUSE, siteId, 15, now)
        Referee.abandon(ses.id)
        val debt = BreakerStore.state.value.abandons.first()
        assertFalse(debt.comboKey.contains("PARTNER"), debt.comboKey)
        assertNotNull(ChallengeEngine.parseCombo(debt.comboKey), "a kulcs visszaolvasható")
    }

    @Test fun `a megbizott levetele probatetel, a vegen az o jelmondataval`() {
        addSite("youtube.com")
        assertTrue(Referee.startPartnerRemoval(now).applied, "megbízott nélkül nincs mit levenni")
        val setup = Referee.setPartner("Anna", now)
        val r = Referee.startPartnerRemoval(now)
        assertFalse(r.applied)
        val ses = r.session!!
        assertTrue(ses.pendingPartnerRemoval)
        assertTrue(ses.steps.last() is Step.Partner)
        solveUntilPartner(ses.id)
        val done = Referee.submitAnswer(ses.id, setup.phrase, now)
        assertTrue(done.sessionDone)
        assertNull(BreakerStore.state.value.partner, "a megbízott lekerült")
        assertEquals(1, BreakerStore.state.value.unlockLog.size, "a lazítás a naplóban")
    }

    @Test fun `ha a megbizott kozben lekerult, a lepese targytalan - atmegy`() {
        val siteId = addSite("youtube.com")
        Referee.setPartner("Anna", now)
        val ses = Referee.startSession(Kind.PAUSE, siteId, 15, now)
        solveUntilPartner(ses.id)
        BreakerStore.mutate { it.copy(partner = null) }
        assertTrue(Referee.submitAnswer(ses.id, "bármi", now).sessionDone)
    }

    @Test fun `a folyamatban levo levetel es a lepes tuleli a mentest`() {
        val toJson = BreakerStore::class.java.getDeclaredMethod("toJson", AppState::class.java).apply { isAccessible = true }
        val fromJson = BreakerStore::class.java.getDeclaredMethod("fromJson", JSONObject::class.java).apply { isAccessible = true }
        Referee.setPartner("Anna", now)
        Referee.startPartnerRemoval(now)
        val saved = toJson.invoke(BreakerStore, BreakerStore.state.value).toString()
        val back = fromJson.invoke(BreakerStore, JSONObject(saved)) as AppState
        assertTrue(back.session!!.pendingPartnerRemoval)
        assertTrue(back.session!!.steps.last() is Step.Partner)
        assertEquals("Anna", (back.session!!.steps.last() as Step.Partner).name)
    }

    @Test fun `a felvetel es a levetel lepteti a blobot, es a jel a blob szama`() {
        var state = AppState()
        assertEquals(state.copy(focusRevFp = SyncRevisions.focusFingerprint(state)), SyncRevisions.bumpFocus(state, "dev", now), "üres állapot: nincs léptetés")
        state = state.copy(partner = PartnerLogic.makeLock("Anna", phrase, now))
        state = SyncRevisions.bumpFocus(state, "dev", now)
        assertEquals(1L, state.focusRev)
        assertEquals(1, state.partnerRev)
        assertEquals(state, SyncRevisions.bumpFocus(state, "dev", now + 1), "változatlan: nem léptet")
        state = SyncRevisions.bumpFocus(state.copy(partner = null), "dev", now + 2)
        assertEquals(2L, state.focusRev, "a levétel is döntés")
        assertEquals(2, state.partnerRev)
        // Egy másik eszközről átvett megbízott: a lenyomat és a kulcs újraszámolva —
        // nincs léptetés, és egy későbbi saját szerkesztés sem bélyegzi át a jelét
        // (azzal a másik eszköz levételét írná felül: azonos jelnél a beállított nyer).
        val adopted = SyncRevisions.adoptFocus(state.copy(partner = PartnerLogic.makeLock("Béla", phrase, now), partnerRev = 7))
        assertEquals(adopted, SyncRevisions.bumpFocus(adopted, "dev", now + 3), "az átvétel nem szerkesztés")
        val edited = SyncRevisions.bumpFocus(
            adopted.copy(focusPacks = listOf(Focus.FocusPack("p1", "Írás", emptyList(), emptyList(), 25))), "dev", now + 4,
        )
        assertEquals(3L, edited.focusRev)
        assertEquals(7, edited.partnerRev, "az átvett megbízott jele marad")
    }
}

File: desktop/test/partner.test.ts
// PÁRBAN ZÁROLÁS: a lazítás végén a megbízott jelmondata is kell.
//
//   UPDATE_PARTNER_FIXTURE=1 npm test   — a fixtures/partner-hash.json újraírása
//
// A fixture a Kotlin és a Swift tükörnek szól: ugyanabból a jelmondatból és
// sóból ugyanazt a lenyomatot kell számolniuk, különben a gépen felvett
// megbízott a telefonon sosem stimmelne — csendben.

import { test } from 'node:test';
import * as assert from 'node:assert/strict';
import * as fs from 'fs';
import * as path from 'path';
import {
  MAX_PARTNER_NAME, MAX_PARTNER_TRIES, mergePartner, normalizePartnerLock, normalizePartnerName,
  normalizePhrase, type PartnerLock,
} from '../src/shared/partner';
import { hashPhrase, makePartnerLock, verifyPhrase } from '../src/helper/partner-crypto';
import { defaultState, newId, type HelperState } from '../src/helper/state';
import * as referee from '../src/helper/referee';
import { adoptFocusRevision, bumpFocusRevision } from '../src/helper/revisions';
import { emptyFocus, mergeFocus, normalizeSyncFocus, sameFocus } from '../src/shared/sync/focus-merge';
import { parseCombo, reverseString, type Step } from '../src/shared/challenges';

const FIXTURE = path.resolve(__dirname, '..', '..', '..', 'fixtures', 'partner-hash.json');
const PHRASE = 'alma bogrács cinege délután';
const SALT = 'AAAAAAAAAAAAAAAAAAAAAA==';

test('a jelmondat kanonikus alakja: kis-nagybetű, szóközök, NFKC nem számít', () => {
  assert.equal(normalizePhrase('  Alma  BOGRÁCS\tcinege\n délután '), PHRASE);
  // Bontott ékezet (a + kombináló vessző) ugyanaz, mint az összetett á.
  assert.equal(normalizePhrase('a' + String.fromCharCode(0x0301) + 'lma bogrács cinege délután'), normalizePhrase('álma bogrács cinege délután'));
  assert.equal(normalizePhrase('   '), '');
  assert.equal(normalizePartnerName('  Anna   Kovács '), 'Anna Kovács');
  assert.equal(normalizePartnerName(''), null);
  assert.equal(normalizePartnerName(42), null);
  assert.equal(normalizePartnerName('x'.repeat(100))!.length, MAX_PARTNER_NAME);
});

test('a lenyomat: a jó jelmondat átmegy bármilyen alakban, a rossz és az üres nem', () => {
  const lock = makePartnerLock('Anna', PHRASE, 1000);
  assert.equal(lock.name, 'Anna');
  assert.ok(lock.salt.length >= 16 && lock.hash.length >= 32);
  assert.ok(verifyPhrase(lock, 'ALMA  bogrács  Cinege délután '));
  assert.ok(!verifyPhrase(lock, 'alma bogrács cinege este'));
  assert.ok(!verifyPhrase(lock, ''));
  // Két felvétel két só: ugyanaz a jelmondat más lenyomat.
  const again = makePartnerLock('Anna', PHRASE, 1000);
  assert.notEqual(again.salt, lock.salt);
  assert.notEqual(again.hash, lock.hash);
  assert.deepEqual(normalizePartnerLock(lock), lock);
  assert.equal(normalizePartnerLock({ ...lock, hash: 'rövid' }), null);
  assert.equal(normalizePartnerLock({ ...lock, name: '' }), null);
  assert.equal(normalizePartnerLock('x'), null);
});

test('a fixtures/partner-hash.json a gép lenyomata — a telefonok ehhez mérik magukat', () => {
  const current = { phrase: PHRASE, salt: SALT, hash: hashPhrase(PHRASE, SALT), wrong: 'alma bogrács cinege este' };
  if (process.env.UPDATE_PARTNER_FIXTURE) {
    fs.writeFileSync(FIXTURE, JSON.stringify(current, null, 2) + '\n');
  }
  assert.ok(fs.existsSync(FIXTURE), `nincs fixture: ${FIXTURE} — UPDATE_PARTNER_FIXTURE=1 npm test`);
  const stored = JSON.parse(fs.readFileSync(FIXTURE, 'utf8'));
  assert.deepEqual(stored, current, 'a fixtures/partner-hash.json elavult — UPDATE_PARTNER_FIXTURE=1 npm test');
  // A lenyomat determinisztikus: ugyanaz a só, ugyanaz a lenyomat.
  assert.equal(hashPhrase(' Alma bogrács CINEGE délután', SALT), stored.hash);
  assert.notEqual(hashPhrase(stored.wrong, SALT), stored.hash);
});

// ------------------------------------------------------------------ a bíró

function stateWithSite(): { state: HelperState; siteId: string } {
  const state = defaultState();
  const siteId = newId('site');
  state.sites.push({
    id: siteId, domain: 'youtube.com', hostnames: ['youtube.com', 'www.youtube.com'],
    addedAt: 1, pauseUntil: null, pendingDeleteAt: null,
  });
  return { state, siteId };
}

function solveStep(step: Step, now: number): string {
  switch (step.type) {
    case 'TRANSCRIBE': return step.text;
    case 'MATH_CHAIN': return String(step.problems[step.pos].a);
    case 'MEMORY': step.armedAt = now - step.showMs - step.waitMs - 1000; return step.code;
    case 'REVERSE': return reverseString(step.text);
    case 'DELAY': throw new Error('a várakozást átvenni kell, nem megválaszolni');
    case 'PARTNER': throw new Error('a megbízott lépése a jelmondat');
  }
}

/** Végigviszi a kísérletet a megbízott lépéséig — a várakozást is átveszi. */
function solveUntilPartner(state: HelperState, sessionId: string, now: number): void {
  let guard = 0;
  while (state.session && guard++ < 200) {
    const step = state.session.steps[state.session.stepIndex];
    if (step.type === 'PARTNER') return;
    if (step.type === 'DELAY') {
      step.claimableAt = now - 1;
      referee.claimDelay(state, sessionId, now);
      continue;
    }
    referee.submitAnswer(state, sessionId, solveStep(step, now), now);
  }
  throw new Error('a kísérlet elfogyott a megbízott lépése előtt');
}

test('megbízottal a terv utolsó lépése az ő jelmondata — a várakozás után', () => {
  const { state, siteId } = stateWithSite();
  const now = 1_700_000_000_000;
  const { phrase, name } = referee.setPartner(state, '  Anna ', now);
  assert.equal(name, 'Anna');
  assert.equal(phrase.split(' ').length, 4, 'négy szó');
  assert.equal(phrase, phrase.toLowerCase());
  assert.ok(state.partner && state.partner.name === 'Anna');
  assert.throws(() => referee.setPartner(state, 'Béla', now), /Már van megbízott/);

  referee.startSession(state, 'pause', siteId, 15, now);
  const steps = state.session!.steps;
  assert.equal(steps[steps.length - 1].type, 'PARTNER');
  assert.equal(steps[steps.length - 2].type, 'DELAY');
  assert.equal((steps[steps.length - 1] as { name: string }).name, 'Anna');

  const id = state.session!.id;
  solveUntilPartner(state, id, now);
  // Rossz jelmondat: nem sorsol újat, csak számol; a plafonnál a kísérlet elszáll.
  for (let i = 1; i < MAX_PARTNER_TRIES; i++) {
    const r = referee.submitAnswer(state, id, 'alma bogrács cinege este', now);
    assert.equal(r.accepted, false);
    assert.ok(r.message && r.message.includes('Nem ez a jelmondat'));
    assert.ok(state.session, 'a kísérlet még él');
  }
  assert.throws(() => referee.submitAnswer(state, id, 'megint rossz', now), /elölről/);
  assert.equal(state.session, null, 'a plafonnál elszállt');
  assert.equal(state.sites[0].pauseUntil, null, 'feloldás nem történt');

  // Újra: minden lépés elölről, és a jó jelmondat a végén feloldja.
  const again = referee.startSession(state, 'pause', siteId, 15, now + 1000);
  solveUntilPartner(state, again.id, now + 1000);
  const r = referee.submitAnswer(state, again.id, ` ${phrase.toUpperCase()} `, now + 1000);
  assert.equal(r.accepted, true);
  assert.equal(r.sessionDone, true);
  assert.ok(state.sites[0].pauseUntil !== null && state.sites[0].pauseUntil > now, 'a szünet elindult');
});

test('a feladott kísérlet kombinációja nem tartalmazza a megbízott lépését', () => {
  const { state, siteId } = stateWithSite();
  const now = 1_700_000_000_000;
  referee.setPartner(state, 'Anna', now);
  const info = referee.startSession(state, 'pause', siteId, 15, now);
  referee.abandonSession(state, info.id);
  const debt = (state.abandons ?? [])[0];
  assert.ok(debt, 'a feladás adóssága megvan');
  assert.ok(!debt.comboKey.includes('PARTNER'), debt.comboKey);
  assert.ok(parseCombo(debt.comboKey), 'a kulcs visszaolvasható — a hűtés alatt ugyanaz jön vissza');
});

test('a megbízott levétele próbatétel, a végén az ő jelmondatával', () => {
  const { state } = stateWithSite();
  const now = 1_700_000_000_000;
  assert.equal(referee.startPartnerRemoval(state, now).applied, true, 'megbízott nélkül nincs mit levenni');
  const { phrase } = referee.setPartner(state, 'Anna', now);
  const r = referee.startPartnerRemoval(state, now);
  assert.equal(r.applied, false);
  assert.ok(r.session);
  assert.equal(state.session!.steps[state.session!.steps.length - 1].type, 'PARTNER');
  solveUntilPartner(state, r.session!.id, now);
  const done = referee.submitAnswer(state, r.session!.id, phrase, now);
  assert.equal(done.sessionDone, true);
  assert.equal(state.partner, undefined, 'a megbízott lekerült');
  assert.equal(state.unlockLog.length, 1, 'a lazítás a naplóban');
});

test('ha a megbízott közben lekerült (szinkron), a lépése tárgytalan: átmegy', () => {
  const { state, siteId } = stateWithSite();
  const now = 1_700_000_000_000;
  referee.setPartner(state, 'Anna', now);
  const info = referee.startSession(state, 'pause', siteId, 15, now);
  solveUntilPartner(state, info.id, now);
  delete state.partner;
  const r = referee.submitAnswer(state, info.id, 'bármi', now);
  assert.equal(r.sessionDone, true);
});

// ------------------------------------------------------------ a jel és a szinkron

test('a felvétel és a levétel lépteti a blobot, és a megbízott jele a blob száma', () => {
  const state = defaultState();
  const now = 1_700_000_000_000;
  assert.equal(bumpFocusRevision(state, 'dev', now), false, 'üres állapot: nincs mit léptetni');
  referee.setPartner(state, 'Anna', now);
  assert.equal(bumpFocusRevision(state, 'dev', now), true);
  assert.equal(state.partnerRev, state.focusRev);
  const rev = state.focusRev!;
  assert.equal(bumpFocusRevision(state, 'dev', now + 1), false, 'változatlan: nem léptet');
  delete state.partner;
  assert.equal(bumpFocusRevision(state, 'dev', now + 2), true, 'a levétel is döntés');
  assert.equal(state.partnerRev, rev + 1);
  // Egy másik eszközről átvett megbízott: a lenyomat és a kulcs újraszámolva —
  // nincs léptetés, és egy későbbi saját szerkesztés sem bélyegzi át a jelét
  // (azzal a másik eszköz levételét írná felül: azonos jelnél a beállított nyer).
  state.partner = makePartnerLock('Béla', 'alma bogrács cinege este', now);
  state.partnerRev = 7;
  adoptFocusRevision(state);
  assert.equal(bumpFocusRevision(state, 'dev', now + 3), false, 'az átvétel nem szerkesztés');
  state.focusPacks = [{ id: 'p1', name: 'Írás', allowSites: [], allowApps: [], defaultMinutes: 25 }];
  assert.equal(bumpFocusRevision(state, 'dev', now + 4), true);
  assert.equal(state.partnerRev, 7, 'az átvett megbízott jele marad');
});

test('fésülés: a jel dönt, azonos jelnél a beállított — és a korábban felvett', () => {
  const a: PartnerLock = { name: 'Anna', salt: 'A'.repeat(24), hash: 'B'.repeat(44), setAt: 100 };
  const b: PartnerLock = { name: 'Béla', salt: 'C'.repeat(24), hash: 'D'.repeat(44), setAt: 200 };
  assert.equal(mergePartner(3, a, 5, undefined), undefined, 'a nagyobb jelű levétel átmegy');
  assert.equal(mergePartner(5, undefined, 3, a), undefined, 'a helyi, nagyobb jelű levétel marad');
  assert.equal(mergePartner(3, a, 3, undefined), a, 'azonos jelnél a beállított nyer');
  assert.equal(mergePartner(3, undefined, 3, b), b);
  assert.equal(mergePartner(3, b, 3, a), a, 'mindkettő beállítva: a korábban felvett');
  assert.equal(mergePartner(2, a, 3, b), b, 'nagyobb jel: a másik megbízott');

  const base = emptyFocus('dev');
  const local = { ...base, partner: a, partnerRev: 3, rev: 3 };
  const removed = { ...base, partnerRev: 5, rev: 5, updatedBy: 'other' };
  const merged = mergeFocus(local, removed);
  assert.equal(merged.partner, undefined);
  assert.equal(merged.partnerRev, 5);
  // A régi kliens blobja (jel nélkül) sosem viszi el a megbízottat.
  const old = { ...base, rev: 9, updatedAt: 999, updatedBy: 'old' };
  assert.deepEqual(mergeFocus(local, old).partner, a);
  assert.ok(!sameFocus(local, { ...local, partner: b }), 'a megbízott cseréje különbség: fel kell tölteni');
  // Kívülről jött adat: a rossz alakú megbízott nincs, a jel a blob rev-jéig.
  const n = normalizeSyncFocus({ partner: { name: 'X', salt: 'rövid', hash: 'rövid' }, partnerRev: 99, rev: 4 }, 'dev');
  assert.equal(n.partner, undefined);
  assert.equal(n.partnerRev, undefined);
  const ok = normalizeSyncFocus({ partner: a, partnerRev: 4, rev: 4 }, 'dev');
  assert.deepEqual(ok.partner, a);
  assert.equal(ok.partnerRev, 4);
});
